### DISCUSSION TITLE
Feature request: Sum multiple prices / sum multiple (price, quantity) pairs

### DISCUSSION THREAD

**Discussion by mrflip** [dinerojs/dinero.js#563] [Ideas]
We would like to be able to call a built-in method sum an iterable of dinero objects, rather than doing sequential binary additions externally -- eg when summing all the line item prices in a purchase order. Part of this is for convenience and pulling all the sensible error checks into the library, but also I get the creepy-crawlies when I see repeated additions of large numbers.

* Would you take a pull request adding a `sum(addends: Iterable<price: Dinero>)` method to the API? 
* Also, a `sumProduct(Iterable<[price: Dinero, multiplier: number]>)` method that would sum the product of each price with the paired multiplier (as you would on a purchase order)?

I believe it is safe to just do these with repeated calls to the outward-facing `add`, `multiply` etc methods, but please give any implementation advice you have to offer. (Illustrating the point that the key virtue of these methods is to take customers of the library from "I _belieeeeeve_ it is safe to just do these with repeated calls" to "I **know** that my calculations are correct, because Dinero is doing the work")

**Comment by sarahdayan** [comment]
Hi @mrflip!

The documentation provides a userland example of [how to implement an `addMany` (or `sum`) function](https://v2.dinerojs.com/docs/api/mutations/add#add-more-than-two-objects). Would this work for you?

Regarding `sumProduct`, it sounds too specific to make sense as a provided function, but if you believe this is a common use case, I'd gladly accept a PR to document a userland solution. What do you think?

**Comment by mrflip** [comment]
Wow the implementation to enable that addMany snippet is elegant. Yes that works for me.

Here's my question on the sumProduct -- is a naive solution at all dangerous, assuming intermediate numbers never leave the safe integer regime? let's say a customer orders 3.4 furlongs of frillery at $4.7770/furlong, 1.79 drams of drambuie at $3.3330/dram, and 19 thimbles at $2/thimble, will

```
function sumProduct(lineitems) {
  const partials = _.map(lineitems, ([multiplicand, muliplier]) => multiplicand.multiply(multiplier))
  return addMany(partials)
}
```
(would be redone as a reduce and native-only)

carry any risk of truncation, or would it be more accurate to carry out the full calculation before re-converting to a dinero object? I know enough to be nervous about all that but not enough to know what is/isn't dangerous.

If it is dangerous, I believe it should be added, but if not, documenting it as a userland function would be the way to go.

**Comment by sarahdayan** [comment]
Hey @mrflip, sorry for the late follow-up!

Your naive `sumProduct` implementation is safe. `multiply` with a scaled amount increases the result's scale to preserve full precision (everything stays as integers internally, no floating-point math involved). Then `add` normalizes scales before summing, so there's no truncation.

The only thing to watch for is amounts exceeding `Number.MAX_SAFE_INTEGER` after multiplication. If that's a concern for your use case, you can use the [`bigint` entry point](https://v2.dinerojs.com/guides/precision-and-large-numbers.html).
